client/core/trade.go

```diff
@@ -1405,60 +1405,95 @@ func (c *Core) tick(t *trackedTrade) (assetMap, error) {
 	t.mtx.Lock()
 	defer t.mtx.Unlock()
 	var swaps, redeems, refunds []*matchTracker
-	assets := make(assetMap)
-	errs := newErrorSet(t.dc.acct.host + " tick: ")
+	assets := make(assetMap) // callers expect non-nil map even on error
 
 	// Check all matches for and resend pending requests as necessary.
 	c.resendPendingRequests(t)
 
 	// Make sure we have a redemption fee suggestion cached.
 	t.cacheRedemptionFeeSuggestion()
 
-	// Check all matches and send swap, redeem or refund as necessary.
 	var sent, quoteSent, received, quoteReceived uint64
-	for _, match := range t.matches {
+	checkMatch := func(match *matchTracker) error {
 		side := match.Side
 		if (side == order.Maker && match.Status >= order.MakerRedeemed) ||
 			(side == order.Taker && match.Status >= order.MatchComplete) {
-			continue
+			return nil
 		}
 		if match.Address == "" {
-			continue // a cancel order match
+			return nil // a cancel order match
 		}
 		if !t.matchIsActive(match) {
-			continue // either refunded or revoked requiring no action on this side of the match
+			return nil // either refunded or revoked requiring no action on this side of the match
 		}
-		switch {
-		case t.isSwappable(c.ctx, match):
+
+		// The trackedTrade mutex is locked, so we must not hang forever. Give
+		// this a generous timeout because it may be necessary to retrieve full
+		// blocks, and catch timeout/shutdown after each check.
+		ctx, cancel := context.WithTimeout(c.ctx, 30*time.Second)
+		defer cancel()
+
+		if t.isSwappable(ctx, match) {
 			c.log.Debugf("Swappable match %s for order %v (%v)", match, t.ID(), side)
 			swaps = append(swaps, match)
 			sent += match.Quantity
 			quoteSent += calc.BaseToQuote(match.Rate, match.Quantity)
+			return nil
+		}
+		if ctx.Err() != nil { // may be here because of timeout or shutdown
+			return ctx.Err()
+		}
```

Comment:
These checks seem unnecessary. The `ctx` will only be **used** in one of these checks, and match metadata checks will preclude rpc calls on the others.

Reply:
> The `ctx` will only be **used** in one of these checks, and match metadata checks will preclude rpc calls on the others

Can you expand on this?  The ctx is passed to each check and a timeout or shutdown can happen at any stage of these checks regardless of which checks actually make use of the ctx.

But it's not obvious or guaranteed that only one of these checks will actually use the ctx.

Reply:
| Method  | Context User | Pre-filters 
| ------------- | ------------- | ------------- |
| `isSwappable`  | `counterPartyConfirms`  | no swapErr or revocation, `Taker` in `MakerSwapCast`  |
| `isSwappable`  | `swapConfirmations`  | no swapErr or revocation, `Maker` in `MakerSwapCast`  |
| `isRedeemable`  | `counterPartyConfirms`  | no swapErr or revocation, `Maker` in `TakerSwapCast`  |
| `shouldBeginFindRedemption`  | `swapConfirmations`  | no search running, `Maker` in `MakerSwapCast`  |

So you're right. It **is** possible to get past the pre-filters of `isSwappable` and `shouldBeginFindRedemption` both on the same `tick`.

Reply:
Wow, thanks for the nice summation table!

But what I'm thinking is that even if you didn't get to any wallet requests for more than one of those *is* checks, I think we'd still want the `ctx.Err()` check after each one, because we don't know in which one(s) there might be a timeout.

Reply:
Haha, that's pretty much **my** argument [here](https://github.com/decred/dcrdex/pull/1751#discussion_r935751867). As the caller, `fetchFiatExchangeRates` doesn't know that `refreshFiatRates` has a timeout, so should use the safer pattern of creating a new timer every tick.

I don't know. You can leave it.

Reply:
Final note on this issue in the following related comment: https://github.com/decred/dcrdex/pull/1732/files#r937073616

A normal pattern would have a function return an error, and if that function took a context that error could be a context error,  however these just have a bool.  A false can mean either it confidently determined that result, or there was simply an error (timeout, shutdown, or other exception).  Instead of refactoring these functions, we at least have the capability to see if the `false` was because of a timeout or shutdown rather than it actually being false **and moving on to the next check**.  Maybe we can refactor this even more in the future so it isn't quite so awkward, but this is the reason.